### PR TITLE
[ckpt] feat: optional local staging for the DCP write

### PR DESCRIPTION
**Pull request by TimYangst**
Add `train.checkpoint.stage_dir`. When set, the DCP write lands under that directory and is copied to `output_dir` afterwards, instead of streaming straight to a destination that may be far slower than local disk.

The motivating failure: a 16-GPU run whose `output_dir` is a FUSE-mounted network filesystem wrote a 302 GiB checkpoint at ~230 MB/s per node, so the save blocked the training loop for 11 minutes. The ranks that were not writing sat in a scalar collective, NCCL's default 600s watchdog fired, and the job was killed and restarted from step 0 -- repeatedly. Writing locally first keeps the blocking phase short enough to stay inside that budget.

The flag is a plain switch, not a policy engine: no directory is probed, no free space is checked, and nothing silently falls back to a direct write. Which scratch device a node exposes, and whether it can hold every local rank's share of the model plus optimizer state, is deployment-specific knowledge the caller already has -- guessing at it here would only hide misconfiguration.

What the checkpointer does own:

- Promotion is done by one rank per node over the whole staging directory, rather than each rank copying the files it believes it wrote, so it does not depend on DCP's file naming.
- `.metadata` is copied last, and only by the coordinator. DCP reads its presence as "this checkpoint is complete", so promoting it earlier would let a reader pick up a half-written checkpoint.
- The staged copy is removed in a `finally`, and a failed `execute_save` cleans up before re-raising: it is as large as the model plus its optimizer state, so leaving it behind would fill the scratch disk for later runs on that node.
- `stage_dir` together with `save_async` is rejected up front, before anything is created on disk. The staged copy is dropped when `save()` returns, which for an async save is before the write has finished.

Default is None, so behaviour is unchanged unless the flag is set.

### What does this PR do?

> Concise overview of the change. Reference related issues/PRs.

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

> Validation results (training curves, eval metrics) for changes not covered by CI.

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

## Summary by CodeRabbit

* **New Features**
  * Added an optional local staging directory for checkpoint saves before copying them to the final destination.
  * Staged checkpoint paths now use deterministic, collision-resistant identifiers.
  * Added validation preventing local staging from being combined with an explicit storage writer.
  * Documented that local staging is incompatible with asynchronous checkpoint saving.

* **Tests**
  * Expanded coverage for staging validation, path handling, cleanup, and distributed checkpoint coordination.

### CODE REVIEW

File: veomni/checkpoint/dcp_checkpointer.py [hunk 1]
```diff
@@ -450,6 +452,56 @@ def restore_extra_parallel_dim(
...
+                os.makedirs(final_path, exist_ok=True)
+                shutil.copyfile(src, os.path.join(final_path, metadata_name))
+    finally:
+        # Always drop the staged copy: it is typically as large as the model plus
+        # its optimizer state, and leaving it behind on a failure would fill the
+        # scratch disk for every later run on that node.
+        if is_node_leader:
+            shutil.rmtree(stage_path, ignore_errors=True)
+        if dist.is_initialized():
+            dist.barrier()
```
[the remark this answers is not in the document]
Comment (Coach257) [reply veomni/checkpoint/dcp_checkpointer.py]: The copy-failure deadlock is a real bug and should be fixed before merge. The coordinator-vs-node-leader `rmtree` race is not, under a normal launcher.

If a node leader raises from `copyfile` before the first barrier, it enters `finally` and waits at the second barrier; every other rank is still on the first. Those collectives no longer match and the save hangs. That is especially plausible here: the destination is the slow FUSE/NFS path this feature exists to avoid. `test_staged_copy_is_removed_even_when_promotion_fails` cannot catch it because it mocks `dist.is_initialized` to `False`.

The second claim does not hold for `torchrun`: rank 0 is `LOCAL_RANK=0` on node 0, so `.metadata` is copied and `stage_path` is removed on the same rank, in that order. Other nodes only delete their own local staging dir. `test_coordinator_creates_the_destination_for_the_metadata` patches `_local_rank` to 1 to isolate the coordinator branch in one process; that is not a production rank map.

Fix: every rank must take the same barrier sequence on success and on copy failure (capture, then re-raise after the collectives). A barrier between the `.metadata` copy and `rmtree` is enough as defense in depth. Please also add a test with `dist.is_initialized=True` that asserts the same number of `barrier` calls when `copyfile` fails.

File: tests/checkpoints/test_dcp_stage_dir.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+"""Unit tests for staging the DCP write on a local filesystem.
```
Comment (Coach257) [inline tests/checkpoints/test_dcp_stage_dir.py]: Please fold this module into `tests/checkpoints/test_dcp_checkpointer.py` instead of adding a new file.

GPU/NPU unit tests invoke that file by path (`gpu_unit_tests.yml` / `npu_unit_tests.yml`). This new file is not on either list, so CI will not run it. The tests are the same style as the existing DCP unit tests (tmp_path + mocked `dist`); `TestPromoteStagedCheckpoint` and `TestStageDirValidation` belong next to the save-planner / async-lifecycle cases already there.

`tests/checkpoints/test_checkpoint_callback.py` only needs the `stage_dir=None` fixture field, which is already in this PR. No workflow change is required once the tests live in `test_dcp_checkpointer.py`.

File: veomni/checkpoint/dcp_checkpointer.py [hunk 3]
```diff
@@ -450,6 +452,56 @@ def restore_extra_parallel_dim(
...
+    is_node_leader = _local_rank() == 0
+    is_coordinator = (not dist.is_initialized()) or dist.get_rank() == 0
+
+    try:
+        if is_node_leader:
+            names = [n for n in sorted(os.listdir(stage_path)) if n != metadata_name]
+            os.makedirs(final_path, exist_ok=True)
+
+            def _copy(name: str) -> None:
+                shutil.copyfile(os.path.join(stage_path, name), os.path.join(final_path, name))
```
[the remark this answers is not in the document]
Comment (Coach257) [reply veomni/checkpoint/dcp_checkpointer.py:530]: The overwrite-while-`.metadata`-is-visible case is real, but it is not a merge blocker for this PR.

The callback writes `global_step_{N}` and only retries a step after a failed save (`_last_saved_step`). A complete checkpoint at the same path is not the normal training path, and the save is synchronous, so there is typically no concurrent reader. Direct DCP writes already overwrite in place without unlinking `.metadata` first; staging does not uniquely introduce that hole.

Unlinking `.metadata` before promoting data is still consistent with the "marker means complete" invariant this PR documents, so I would take it as cheap hardening, not as a requirement to land.

File: veomni/checkpoint/dcp_checkpointer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/checkpoint/dcp_checkpointer.py]: Good catch,  fixed

File: veomni/checkpoint/dcp_checkpointer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/checkpoint/dcp_checkpointer.py:530]: Agreed, and fixed even though it predates this PR.

File: tests/checkpoints/test_dcp_stage_dir.py (hunk 2 above)
Comment (TimYangst) [reply tests/checkpoints/test_dcp_stage_dir.py]: Merged into tests/checkpoints/test_dcp_checkpointer.py and deleted the separate file.

File: tests/checkpoints/test_dcp_checkpointer.py [hunk 4]
```diff
@@ -1108,3 +1108,203 @@ def test_missing_extra_state_key_load(self, mock_dist, tmp_path):
...
+    def test_every_rank_runs_the_same_barriers_when_a_copy_fails(self, staged, failing_role):
+        """Collective parity: barriers are untagged, so one rank skipping one hangs the job.
+
+        Whichever role fails, every role must reach the same number of barriers and
+        surface the error only after the last one.
+        """
+        from veomni.checkpoint.dcp_checkpointer import _promote_staged_checkpoint
+
+        counts = {}
+        for role, (global_rank, local_rank) in self._ROLES.items():
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/checkpoints/test_dcp_checkpointer.py:1239]: Fixed — the fixture is a factory now, so each role gets its own staged copy.

File: tests/checkpoints/test_dcp_checkpointer.py [hunk 5]
```diff
@@ -1108,3 +1108,203 @@ def test_missing_extra_state_key_load(self, mock_dist, tmp_path):
...
+            stage_path, final_path = staged
+            barrier = MagicMock()
+            with patch("veomni.checkpoint.dcp_checkpointer.dist.is_initialized", return_value=True):
+                with patch("veomni.checkpoint.dcp_checkpointer.dist.barrier", barrier):
+                    with patch("veomni.checkpoint.dcp_checkpointer.dist.get_rank", return_value=global_rank):
+                        with patch("veomni.checkpoint.dcp_checkpointer._local_rank", return_value=local_rank):
+                            with patch(
+                                "veomni.checkpoint.dcp_checkpointer.shutil.copyfile",
+                                side_effect=OSError("boom") if role == failing_role else None,
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/checkpoints/test_dcp_checkpointer.py:1249]: Right, that case asserted nothing. Failure injection is limited to the two copying roles now; the participant still takes part in the barrier-count comparison, which is why it's there.

File: tests/checkpoints/test_dcp_checkpointer.py [hunk 6]
```diff
@@ -1108,3 +1108,203 @@ def test_missing_extra_state_key_load(self, mock_dist, tmp_path):
...
+
+    def test_stage_key_does_not_collide_across_similar_paths(self):
+        """Separator substitution maps /tmp/a_b/c and /tmp/a/b_c onto one directory."""
+        from veomni.checkpoint.dcp_checkpointer import _stage_key
+
+        assert _stage_key("/tmp/a_b/c") != _stage_key("/tmp/a/b_c")
+        assert _stage_key("/tmp/run/step_1") == _stage_key("/tmp/run/step_1")
+        assert _stage_key("/tmp/run/step_1") != _stage_key("/tmp/run/step_2")
+        # relative and absolute spellings of one destination stage together
+        assert _stage_key("/tmp/run/step_1") == _stage_key("/tmp/run/./step_1")
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply tests/checkpoints/test_dcp_checkpointer.py]: Fixed — compares a relative path against its absolute spelling now.

**Review by Coach257 — APPROVED** [review]
